### PR TITLE
Data out sstream

### PR DESCRIPTION
**Pull request by RebeccaGibbon**
Writing data to the yaml file using stringstream in the class deconstructor.

### CODE REVIEW

File: my_2d_robot/CMakeLists.txt [hunk 1]
```diff
@@ -103,10 +108,10 @@ find_package(catkin REQUIRED COMPONENTS
 ## CATKIN_DEPENDS: catkin_packages dependent projects also need
 ## DEPENDS: system dependencies of this project that dependent projects also need
 catkin_package(
-# INCLUDE_DIRS include
+ INCLUDE_DIRS include
```
Comment (chris24sahadeo) [inline my_2d_robot/CMakeLists.txt]: Looks like you uncommented this line during the rebase.
This now looks for an "include" dir in the package (of which you have none) causing the build to fail.

Two solutions:
1. (Do this for now) Re-comment the line and the build will pass.
2. (Do this for your next milestone) Make a header file `include/joint_states_subscriber.h` and prototype your `SensorMeasurementData` class there instead. This is a good C++ practice to follow anyway :).

File: my_2d_robot/CMakeLists.txt [hunk 2]
```diff
@@ -133,6 +138,7 @@ include_directories(
 ## Declare a C++ executable
 ## With catkin_make all packages are built within a single CMake context
 ## The recommended prefix ensures that target names across packages don't collide
+#add_executable(${PROJECT_NAME}_node src/my_2d_robo_state_publisher.cpp)
```
Comment (chris24sahadeo) [inline my_2d_robot/CMakeLists.txt]: Remove redundant line.

File: my_2d_robot/CMakeLists.txt [hunk 3]
```diff
@@ -148,14 +154,20 @@ include_directories(
 # target_link_libraries(${PROJECT_NAME}_node
 #   ${catkin_LIBRARIES}
 # )
+#For publisher:
+#target_link_libraries(${PROJECT_NAME}_node ${catkin_LIBRARIES})
```
Comment (chris24sahadeo) [inline my_2d_robot/CMakeLists.txt]: Remove redundant lines.

File: my_2d_robot/CMakeLists.txt [hunk 4]
```diff
@@ -148,14 +154,20 @@ include_directories(
...
 
 # PUBLISHER NODE 
 add_executable(${PROJECT_NAME}_node src/my_2d_robo_state_publisher.cpp)
 target_link_libraries(${PROJECT_NAME}_node ${catkin_LIBRARIES})
 
-# SUBSCRIBER NODE
+
+#SUBSCRIBER NODE
```
Comment (chris24sahadeo) [inline my_2d_robot/CMakeLists.txt]: Add space between # and actual comment. Parse the file for any similar errors that I might've missed.

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 5]
```diff
@@ -1,19 +1,150 @@
...
+#include <eigen3/Eigen/Dense>
 
-void joint_statesCallback(const std_msgs::String::ConstPtr& msg)
+using namespace Eigen;
+typedef Matrix<float, 2, 1> Vector2f;
+
+class SensorMeasurementData
+{
+public:
+  // Contructor.
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp]: You can omitt obvious comments. Same goes for destructor.

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 6]
```diff
@@ -1,19 +1,150 @@
...
+#include <sensor_msgs/JointState.h>
+#include <string>
+#include <fstream>
+#include <sstream>
+#include <boost/filesystem.hpp>
+#include "yaml-cpp/yaml.h"
+#include <eigen3/Eigen/Dense>
 
-void joint_statesCallback(const std_msgs::String::ConstPtr& msg)
+using namespace Eigen;
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp]: We should never use "using namespace X" since it messes with scoping.

Instead follow the guide: http://wiki.ros.org/CppStyleGuide#Namespaces-1

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 7]
```diff
@@ -1,19 +1,150 @@
...
+typedef Matrix<float, 2, 1> Vector2f;
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp]: Good thinking with the aliasing, but Eigen already has a built in `Vector2d` class - so we should use that instead. Also use doubles over floats for more precision.

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 8]
```diff
@@ -1,19 +1,150 @@
...
+  position_joint2_ = msg->position[1];
+
+  ee_position_ = eePos(position_joint1_, position_joint2_);
+
+  saveYaml();
 }
 
-int main (int argc, char **argv){
+// Method to write the joint state angles and end effector position to yaml file.
+void SensorMeasurementData::saveYaml()
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp]: This function does not actually do the saving - function names should describe what they do as precisely as possible. Instead something like `save_joint_angle_eepos`

Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp]: Your current output looks like:
```yaml
joint1 angle: 334
joint2 angle: 199
end effector x position: -0.5908343
end effector y position: -1.065893
```
Although you are using the YAML library, this is not actual yaml formatted output. The file should look be a sequence of mapping of sequences:

```yaml
-
  theta: [0, 3.14]
  ee_pos: [0, 1]
- 
  theta: [0, 3.14]
  ee_pos: [0, 1]
```

See some valid YAML formats here: https://yaml.org/spec/1.2/spec.html#Introduction

Rethinking, it might actually be easier to write output directly to the opened file as soon as you calculate an eepos.

Consider the pseudocode workflow:

```
open new file in constructor

close file in destructor

in function callback:
eepos = calc_eepos()
write directly to file here 
```

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 9]
```diff
@@ -1,19 +1,150 @@
...
+    n->getParam("Theta2_offset", theta2_offset_);
+  }
+
+  // Deconstructor.
+  ~SensorMeasurementData()
+  {
+    std::stringstream stream;
+
+    std::ofstream fout;
+    fout.open(my_output_, std::fstream::app);
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp]: Looks like we keep appending to the same file, when really we want to create a new file on every run. To achieve this you can create a new file with the current unix timestamp appended to the end of the filename. That way - every roslaunch created a new file so we can test new configs

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 10]
```diff
@@ -1,19 +1,150 @@
...
+    position << (link_1_ * cos(Theta1 * M_PI / 180)) + (link_2_ * cos((Theta1 + Theta2) * M_PI / 180)),
+        (link_1_ * sin(Theta1 * M_PI / 180)) + (link_2_ * sin((Theta1 + Theta2) * M_PI / 180));
+
+    return position;
+  }
+};
+
+// Class methods.
+// Method to get the joint state information from publisher node.
+void SensorMeasurementData::joint_statesCallback(const sensor_msgs::JointState::ConstPtr& msg)
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp]: Function names should be camel case.

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 11]
```diff
@@ -1,19 +1,150 @@
...
+typedef Matrix<float, 2, 1> Vector2f;
+
+class SensorMeasurementData
+{
+public:
+  // Contructor.
+  SensorMeasurementData(ros::NodeHandle* n)
+  {
+    // Initialises length of links (specified in urdf).
+    link_1_ = 5.0;
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp]: Values should not be hardcoded - it should work regardless of the 2d model you have in the URDF.

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 12]
```diff
@@ -1,19 +1,150 @@
...
+  float position_joint1_;
+  float position_joint2_;
+  // Joint angle offsets.
+  float theta1_offset_;
+  float theta2_offset_;
+  // Link lengths.
+  float link_1_;
+  float link_2_;
+  // Vector to store the end effector position.
+  Vector2f ee_position_;
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp]: The class does not need to store this state since it changes per joint angle.

File: my_2d_robot/src/my_2d_robo_state_publisher.cpp [hunk 13]
```diff
@@ -23,22 +22,23 @@ int main(int argc, char** argv)
   while (ros::ok)
   {
     joint_state.header.stamp = ros::Time::now();
-    // There is one message per update. This is 2 since 2 joints are updated at a time.
 
-    // Randomly generating values in the range [0,2*PI].
-    float joint1_pos = ((rand() % (2 * 315)) / 100.0);
-    float joint2_pos = ((rand() % (2 * 315)) / 100.0);
+    // Randomly generates angle values in the range [0,360] degrees.
+    float joint1_pos = (rand() % 361);
```
Comment (chris24sahadeo) [inline my_2d_robot/src/my_2d_robo_state_publisher.cpp]: Why the change from rad to deg?

File: my_2d_robot/urdf/my2drobo.urdf [hunk 14]
```diff
@@ -45,6 +44,7 @@
 
     <link name="endlink">
     
+
```
Comment (chris24sahadeo) [inline my_2d_robot/urdf/my2drobo.urdf]: Remove random extra line.

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 15]
```diff
@@ -1,19 +1,150 @@
...
+// Main function.
+int main(int argc, char** argv)
+{
+  ros::init(argc, argv, "joint_states_subscriber");
+  ros::NodeHandle n;
 
-    ros::spin();
+  SensorMeasurementData my_2d_robo_states = SensorMeasurementData(&n);
 
+  ros::spin();
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp:14]: Correct me if I'm wrong. But it looks like your subscriber spins forever and only terminates on a Ctrl+C.

We should be able to specify a certain number of data points that we want in the launch file, your subscriber reads this in from the rosparam, and maintains an internal count in a class variable. When count == num_data points then the subscriber should immediately terminate. This will make data gen easier and is good design.

File: my_2d_robot/src/joint_states_subscriber.cpp (hunk 8 above)
Comment (RebeccaGibbon) [reply my_2d_robot/src/joint_states_subscriber.cpp]: I don't know how to open a file, write to a file and closing the file in different functions. :( 
I spent some time researching this today but I'm still not sure since the fstream object would be implicitly deleted at the end of each function.

File: my_2d_robot/src/joint_states_subscriber.cpp [hunk 16]
```diff
@@ -1,142 +1,9 @@
+#include "joint_states_subscriber.h"
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.cpp:1]: Super minor nit:
Looks like you refactored your code into a header file, which is great! However I think the convention you are using is a bit odd.

This is the typical convention:
```
- my_2d_robot/
  - include/
    - joint_states_subscriber.h  // Classes, variables, function signatures
  - src/
    - joint_state_subscriber.cpp  // Function definitions
    - main.cpp  // includes joint_states_subscriber.h  and uses the functions within
```

But you are defining the functions in your header file and calling them in your .cpp file.

LMK if you understand what I mean, else we can have a chat :)

We can leave it as is for now. Add a `// TODO(rebecca)` comment in your code so that we don't forget to do it later

File: my_2d_robot/src/joint_states_subscriber.h [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+    my_output_.append("_output.yaml");
+  }
+
+  ~SensorMeasurementData()
+  {
+    // fout.close();
+  }
+
+private:
+  ros::Subscriber sub_;
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.h:38]: Move all variable declarations to the end (after functions).

File: my_2d_robot/src/joint_states_subscriber.h [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+  int data_num_max_;
+
+  // Callback method to get joint angles.
+  void jointStatesCallback(const sensor_msgs::JointState::ConstPtr& msg);
+
+  // Method to write the joint state angles and end effector position to yaml file.
+  std::string saveJointAnglesEepos(Vector2d end_effector_position);
+
+  // Method to calculate the position of the end effector.
+  Vector2d eePos(double joint_1, double joint_2)
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.h]: Looks like this is the only method that you defined in the class itself. All the others are defined outside the class and are scoped by the class. Any particular reason for this?

If not then we can move the method declaration outside the class for the sake of uniformity.

Same goes for any other methods (Ex the constructor)

File: my_2d_robot/src/joint_states_subscriber.h [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+#include <sstream>
+#include <cmath>
+#include <ctime>
+
+using Eigen::Vector2d;
+
+class SensorMeasurementData
+{
+public:
+  SensorMeasurementData(ros::NodeHandle* n)
```
Comment (chris24sahadeo) [inline my_2d_robot/src/joint_states_subscriber.h]: Make param const.

Most params will be a const ref (&) or const ptr (*)

File: my_2d_robot/CMakeLists.txt [hunk 20]
```diff
@@ -103,9 +108,9 @@ find_package(catkin REQUIRED COMPONENTS
 ## CATKIN_DEPENDS: catkin_packages dependent projects also need
 ## DEPENDS: system dependencies of this project that dependent projects also need
 catkin_package(
-# INCLUDE_DIRS include
+ INCLUDE_DIRS include/${HEADER_FILES}
```
Comment (chris24sahadeo) [inline my_2d_robot/CMakeLists.txt]: Looks like this still needs to be commented. Although you now have header files, they are not actually in an `include` directory. (See my other related comment further below)

File: my_2d_robot/robo2d.launch [hunk 21]
```diff
@@ -2,7 +2,13 @@
         <param name="robot_description" command="cat $(find my_2d_robot)/urdf/my2drobo.urdf" />
         <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
         <node name="my_2d_robo_state_publisher" pkg="my_2d_robot" type="my_2d_robot_node" />
+        <node name="joint_states_subscriber" pkg="my_2d_robot" type="my_2d_robot_subscriber_node" />
+        <param name="Theta1_offset" type="double" value ="10.0" />
+        <param name="Theta2_offset" type="double" value ="5.0" />  
+        <param name="link_1" type="double" value ="5.0" /> 
+        <param name="link_2" type="double" value ="5.0" /> 
         <param name="loop_rate" type="double" value="2.0" />
```
Comment (chris24sahadeo) [inline my_2d_robot/robo2d.launch]: For some reason, I could not speed up the simulation on my end. LMK if you can obtain higher than 2Hz on your laptop.

**Review by chris24sahadeo — APPROVED** [review]
I mentioned a few design points to consider. Otherwise LGTM.

Make changes before merging in :)

File: my_2d_robot/src/joint_states_subscriber.cpp (hunk 16 above)
Comment (RebeccaGibbon) [reply my_2d_robot/src/joint_states_subscriber.cpp:1]: I think I understand. 
So I need to:
- move the function definitions to the joint_state_subscriber.cpp file and 
- create a main.cpp for this node where all the function calls are done? 

However, since the callback function calls the functions to calculate the end effector position and to save the end effector, then the only function call in the main.cpp file will be the one for the constructor, similar to the current joint_states_subscriber.cpp?

File: my_2d_robot/robo2d.launch (hunk 21 above)
Comment (RebeccaGibbon) [reply my_2d_robot/robo2d.launch]: Hmmm, I can't obtain anything higher or lower.
Will look into fixing this.

File: my_2d_robot/src/joint_states_subscriber.cpp (hunk 16 above)
Comment (chris24sahadeo) [reply my_2d_robot/src/joint_states_subscriber.cpp:1]: Precisely!

File: my_2d_robot/robo2d.launch (hunk 21 above)
Comment (chris24sahadeo) [reply my_2d_robot/robo2d.launch]: Could be fixed in a separate PR, immediately after this one.
